codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -474,7 +503,15 @@ public void write(ChannelHandlerContext ctx, int allowedBytes) {
             // Writing headers may fail during the encode state if they violate HPACK limits.
             Throwable failureCause = f.cause();
             if (failureCause == null) {
+                // This is best-effort as the operation is async it may still fail later on. We just update it here
+                // directly as otherwise it may be challenging for the end-user to co-ordinate writes that depend
+                // on the state.
                 stream.headersSent(isInformational);
+
+                if (!f.isSuccess()) {
```

Comment:
can you include a unit test which exercises this code path? I'm not sure this change is necessary.

`FlowControlledHeaders` adds itself as listener to `promise` before the `write*` operation. When the future completes `FlowControlledBase#operationComplete` will be called, if the future has failed `error(..)` will be called, which should call `lifecycleManager.onError(ctx, cause)`. If anything maybe we should remove the `else` condition below which also calls `lifecycleManager.onError(ctx, cause)`?

Reply:
@Scottmitch good catch. Let me remove it for the reason you stated.